### PR TITLE
[distributed] add TrackTime, CUDATrackTime to monitor perf for weight loading per stage and future perf measurements

### PR DESCRIPTION
**Pull request by lessw2020**
This PR:
1 - adds a TrackTimer class to wrap perf_counter and make it easy to do perf timings.
2 - implements it to track weight loading time per stage:

Screenshot 2024-09-08 at 6 22 53 PM

Note - there is a similar timer class called measure_time in chat, but I added some additional rounding options,  unit options and also didn't want to use print for the logging.

### CODE REVIEW

**Comment by kwen2501** [comment]
Does our weight loading include any CUDA ops? 
It seems this timer captures time on the CPU side.

**Comment by lessw2020** [comment]
> Does our weight loading include any CUDA ops? It seems this timer captures time on the CPU side.

there's no express cuda ops beyond a .to from cpu to gpu, but that is a sync event.

That said, I agree we want to have cuda events for timing most everything else, so adding a cuda even timer class.  
I guess I'll do a sep PR for that.

**Comment by lessw2020** [comment]
nvm, makes sense to just add it here.
Added a CUDATrackTime, exact same interface so either drops in with same usage pattern, but CUDATrackTime uses cuda events for potentially more precise timing for any underlying cuda kernel calls.

File: distributed/utils.py [hunk 1]
```diff
@@ -76,3 +81,80 @@ class NoColor:
...
+        if self.active:
+            raise RuntimeError("Timer is already running. Use .stop() to stop it")
+        self.start_event.record()
+        self.active = True
+
+    def stop(self):
+        if not self.active:
+            raise RuntimeError("Timer is not running. Use .start() to start it")
+        self.end_event.record()
+        self.active = False
```
Comment (kwen2501) [inline distributed/utils.py:136]: nit: can you add a note here that these methods (containing `.record()`) can create CUDA context on `current_device()` (whatever it is) if there hasn't been one? Thus for safety, `torch.cuda.set_device()` is strongly needed for `CUDATrackTime`.

File: distributed/utils.py [hunk 2]
```diff
@@ -76,3 +81,80 @@ class NoColor:
...
+
+    def get_time(self):
+        if self.active:
+            raise RuntimeError("Timer is still running. Use .stop() to stop it")
+
+        torch.cuda.synchronize()  # Wait for all GPU operations to finish
```
Comment (kwen2501) [inline distributed/utils.py:142]: nit: do you mind adding a note to `CUDATrackTime` that this timer is going to trigger device synchronization? Thanks!

**Review by kwen2501 — APPROVED** [review]
LGTM. Suggested some notes as the utils themselves make some CUDA calls (which can lead to unexpected effects if users are not aware)

File: distributed/utils.py (hunk 1 above)
Comment (lessw2020) [reply distributed/utils.py:151]: I've updated this to take a device and then to use that for setting the events.  
Also added comment re: creating context on the designated device.

File: distributed/utils.py (hunk 2 above)
Comment (lessw2020) [reply distributed/utils.py]: added comment in doc string re: synchronization.

**Comment by kwen2501** [comment]
The new changes look good! Thanks!

**Review by kwen2501 — APPROVED** [review]

**Pull request merged**